File: 2026/leetcode/23.merge-k-sorted-lists.go
/*
 * @lc app=leetcode.cn id=23 lang=golang
 * @lcpr version=30204
 *
 * [23] 合并 K 个升序链表
 */

// @lcpr-template-start

// @lcpr-template-end
// @lc code=start
// 核心思路：分治合并 K 个有序链表，把区间拆成左右两半分别合并，再合并两个有序链表。
// 关键点：递归边界处理空区间和单链表区间，合并阶段复用原链表节点。
// 复杂度：时间 O(N log k)，空间 O(log k) 递归栈，N 为所有节点总数。
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeKLists(lists []*ListNode) *ListNode {
	return mergeK1Lists(lists, 0, len(lists)-1)
}

func mergeK1Lists(lists []*ListNode, start, end int) *ListNode {
	if start == end {
		return lists[start]
	}
	if start > end {
		return nil
	}
	mid := (start + end) / 2
	left := mergeK1Lists(lists, start, mid)
	right := mergeK1Lists(lists, mid+1, end)
	res := &ListNode{}
	tmp := res
	for left != nil && right != nil {
		if left.Val < right.Val {
			tmp.Next = left
			left = left.Next
		} else {
			tmp.Next = right
			right = right.Next
		}
		tmp = tmp.Next
	}
	if left != nil {
		tmp.Next = left
	}
	if right != nil {
		tmp.Next = right
	}
	return res.Next
}

// @lc code=end

/*
// @lcpr case=start
// [[1,4,5],[1,3,4],[2,6]]\n
// @lcpr case=end

// @lcpr case=start
// []\n
// @lcpr case=end

// @lcpr case=start
// [[]]\n
// @lcpr case=end

*/

File: 2026/leetcode/23.merge-k-sorted-lists.java
/*
 * @lc app=leetcode.cn id=23 lang=java
 * @lcpr version=30204
 *
 * [23] 合并 K 个升序链表
 */


// @lcpr-template-start

// @lcpr-template-end
// @lc code=start

import java.util.List;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
// 方法一： 暴力破解 - 每次取同一个位置最小的
// class Solution {
//     public ListNode mergeKLists(ListNode[] lists) {
//         ListNode res = new ListNode();
//         ListNode tmp = res ;
//         while(true){
//             ListNode min = getMin(lists);
//             if(min == null){
//                 break;
//             }
//             tmp.next = min;
//             tmp = tmp.next;
//         }  
//         return res.next;
//     }

//     ListNode getMin(ListNode[] lists){
//         ListNode min = null;
//         int index = -1;
//         for(int i = 0; i < lists.length; i++){
//             ListNode node = lists[i];
//             if(node != null && (min == null || node.val < min.val)){
//                 min = node;
//                 index = i;
//             }
//         }
//         if(min != null){
//             lists[index] = lists[index].next;
//         }
//         return min;
//     }
// }
// 方法二：暴力破解，逐个合并
// class Solution {
//     public ListNode mergeKLists(ListNode[] lists) {
//         if (lists == null || lists.length == 0) {
//             return null;
//         }
//         return mergeK1Lists(lists, 0);
//     }

//     ListNode mergeK1Lists(ListNode[] lists, int start ){
//         if(start == lists.length-1){
//             return lists[start];
//         }
//         ListNode other = mergeK1Lists(lists, start+1);
//         ListNode current = lists[start];
//         ListNode res = new ListNode();
//         ListNode p = res;
//         while (other != null && current != null) {
//             if(other.val < current.val){
//                 p.next = other;
//                 other = other.next;
//             }else{
//                 p.next = current;
//                 current = current.next;
//             }
//             p = p.next;
//         }
//         if(current != null){
//             p.next = current;
//         }
//         if(other != null){
//             p.next = other;
//         }
//         return res.next;
//     }
// }
// 方法三： 分治
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) {
            return null;
        }
        return mergeK1Lists(lists, 0, lists.length-1);
    }

    ListNode mergeK1Lists(ListNode[] lists, int start, int end  ){
        if(start == end){
            return lists[start];
        }
        if(start > end){
            return null;
        }
        int mid = (start + end)/2;
        ListNode left = mergeK1Lists(lists, start, mid);
        ListNode right = mergeK1Lists(lists, mid+1, end);
        ListNode res = new ListNode();
        ListNode p = res;
        while (left != null && right != null) {
            if(left.val < right.val){
                p.next = left;
                left = left.next;
            }else{
                p.next = right;
                right = right.next;
            }
            p = p.next;
        }
        if(left != null){
            p.next = left;
        }
        if(right != null){
            p.next = right;
        }
        return res.next;
    }
}
// @lc code=end

/*
// @lcpr case=start
// [[1,4,5],[1,3,4],[2,6]]\n
// @lcpr case=end

// @lcpr case=start
// []\n
// @lcpr case=end

// @lcpr case=start
// [[]]\n
// @lcpr case=end

 */

